### PR TITLE
Haxe template

### PR DESCRIPTION
**Pull request by markknol**
Based on <http://old.haxe.org/doc/cross/template>

@Merelleya Can you double check grammar/spelling? :+1: 

Would it also be an idea to mention other template libs?

### CODE REVIEW

File: 10-std.tex [hunk 1]
```diff
@@ -368,6 +368,98 @@ \section{Lambda}
 }
 \end{lstlisting} 
 
+\section{Template}
+\label{std-template}
+
+Haxe comes up with a standard Template System with an easy to use syntax which is interpreted by a lightweight class called \type{haxe.Template}.
```
Comment (Merelleya) [inline 10-std.tex]: Haxe comes with...template system (no capitals)

File: 10-std.tex [hunk 2]
```diff
@@ -368,6 +368,98 @@ \section{Lambda}
...
+
+A Template is a string or a file that is used to produce some HTML or XML output depending on the input. Here is a small template example:
```
Comment (Merelleya) [inline 10-std.tex]: template (no capitals)

File: 10-std.tex [hunk 3]
```diff
@@ -368,6 +368,98 @@ \section{Lambda}
...
+
+\haxe{assets/Template.hx}
+
+The console will trace \ic{My name is Mark, 30 years old}.
+
+\paragraph{Expressions}
+Expression can be put between the \ic{::}, and the syntax allows the current possibilities:
```
Comment (Merelleya) [inline 10-std.tex]: "Expressions" or "an expression", no comma before and. Delete the and, keep the comma

File: 10-std.tex (hunk 1 above)
Comment (markknol) [inline 10-std.tex]: Yes, that was from the old docs. Personally I think it adds quite heavy amount of stuff. I think Ill just remove _lightweight._

Comment (Simn) [inline 10-std.tex]: Lightweight is okay there, it carries the information that the implementation is quite basic and people might want to consider a different/real template implementation for more serious stuff.

File: 10-std.tex (hunk 2 above)
Comment (Simn) [inline 10-std.tex]: Also there's no reason to assume HTML/XML. You can generate any kind of string with haxe.Template.

File: assets/Template.hx [hunk 4]
```diff
@@ -0,0 +1,9 @@
+class Main {
+    static function main() {
+		var sample = "My name is <strong>::name::</strong>, <em>::age::</em> years old";
+		var user = {name:"Mark", age:30};
+		var template = new haxe.Template(sample);
```
Comment (Simn) [inline assets/Template.hx]: indent

In general please use two-space indents for code samples. I realize that's not typical Haxe indenting but we have to be careful with the width of code samples.

File: assets/TemplateMacros.hx [hunk 5]
```diff
@@ -0,0 +1,18 @@
+class Main {
+    static function main() {
+		new Main();
+	}
+	
+    public function new() {
+		var user = {name:"Mark", distance:3500};
+        var sample = "The results: $$display(::user::,::time::)";
```
Comment (Simn) [inline assets/TemplateMacros.hx]: indent

File: assets/TemplateResource.hx [hunk 6]
```diff
@@ -0,0 +1,9 @@
+class Main {
+    static function main() {
+		var sample = haxe.Resource.getString("my_sample");
+		var user = {name:"Mark", age:30};
+		var template = new haxe.Template(sample);
```
Comment (Simn) [inline assets/TemplateResource.hx]: indent

**Comment by markknol** [comment]
Maybe a lazy suggestion but cant we replace `\t` with `\s\s`?

**Comment by Simn** [comment]
We could but it's nicer to see the actual print margin (80) in the editor directly.

**Comment by markknol** [comment]
ok ok ok :smile:

File: 10-std.tex [hunk 7]
```diff
@@ -368,6 +368,98 @@ \section{Lambda}
...
+\end{description}
+
+\paragraph{Conditions}
+It is possible to test conditions using \ic{::if flag1::}. Optionally, the condition may be followed by \ic{::elseif flag2::} or \ic{::else::}. Close the condition with \ic{::end::}.
+
+\begin{lstlisting} 
+::if isValid:: valid ::else:: invalid ::end::
+\end{lstlisting} 
+
+Operators can be used, but it doesn't deal with operator precedence. Therefore it is required to enclose each operation in parentheses \ic{()}. The current operators are allowed: \ic{+}, \ic{-}, \ic{*}, \ic{/}, \ic{>}, \ic{<},  \ic{>=}, \ic{<=}, \ic{==}, \ic{!=}, \ic{&&} and \ic{||}.
```
Comment (Merelleya) [inline 10-std.tex]: no comma before "but", also "it" refers to what?
"Currently, the following operators are allowed:"

File: 10-std.tex [hunk 8]
```diff
@@ -368,6 +368,98 @@ \section{Lambda}
...
+		<tr>
+			<td>::name::</td>
+			<td>::age::</td>
+		</tr>
+	::end::
+</table>
+\end{lstlisting} 
+
+\paragraph{Sub-templates}
+To include templates inside other templates, pass the sub-template result string as a parameter.
```
Comment (Merelleya) [inline 10-std.tex]: include already contains inside. So use "to include templates in other"

File: 10-std.tex [hunk 9]
```diff
@@ -368,6 +368,98 @@ \section{Lambda}
...
+var userOutput = userTemplate.execute({users: users});
+
+var template = new haxe.Template("The users are ::users::");
+var output = template.execute({users: userOutput});
+trace(output);
+\end{lstlisting} 
+The console will trace \ic{The users are Mark(30) John(45)}.
+
+\paragraph{Macros}
+To callback code while parts of the template are rendered, providing a \expr{macros} object to the argument of \expr{Template.execute}. The key will act the template variable name, the value refers to a callback function that should return a \type{String}. The first argument of this macro function is always a \expr{resolve()} method, followed by the given arguments. The resolve function can be called to retreive values from the Template context. If \expr{macros} has no such field, the result is unspecified.
```
Comment (Merelleya) [inline 10-std.tex]: "are being rendered", "provide a". "The key will act as...", "retrieve" (typo), "template" (no caps)

File: 10-std.tex [hunk 10]
```diff
@@ -368,6 +368,98 @@ \section{Lambda}
...
+
+The following example passes itself as macro function context, and executes \ic{display} from the template.
```
Comment (Merelleya) [inline 10-std.tex]: no comma before and

File: 10-std.tex [hunk 11]
```diff
@@ -368,6 +368,98 @@ \section{Lambda}
...
+\haxe{assets/TemplateMacros.hx}
+The console will trace \ic{The results: Mark ran 3.5 kilometers in 15 minutes}.
+
+\paragraph{Globals}
+Use the \expr{Template.globals} object to store those values that should be applied across all Template instances. This has lower priority than the context argument of \expr{Template.execute}.
```
Comment (Merelleya) [inline 10-std.tex]: "the values" instead of "those" (those implies distance). "template instances" (no caps)

**Comment by Simn** [comment]
I thought this was reviewed already. @markknol: Could you send another PR with the fixes?

**Comment by markknol** [comment]
Ah I thought I did a near-perfect job.. :stuck_out_tongue_winking_eye: Oh well.. 

Yes, Ill PR soon :+1:

**Comment by Merelleya** [comment]
I'm so sorry...I just didn't get around to it >.<

**Pull request merged**